File: test/conformance/dart/rows/P3.dart
// test/conformance/dart/rows/P3.dart — row dart/P3, "hostile bytes, sweep +
// sanitizer" (docs/roadmap.sexp:3956; audit schema#898, matrix schema#876).
//
// THE LAW. docs/FIXED-FORM-ALGORITHM.md:1686, the fixed form's proof table
// (§7 proof 5), and its ruling in docs/SPEC-TABLES.md §3.4
// (docs/SPEC-TABLES.md:7138):
//
//   "a byte-flip fuzz over the whole file, under a sanitizer, and it is not
//    optional — every offset is arithmetic over sizes a stranger wrote down,
//    so every byte, one bit at a time, is answered one of three ways and never
//    a fourth: a refusal by name, a `malformed` read, or a read that lands
//    values."
//
// and docs/SPEC-TABLES.md §3.4's first sentence
// (docs/SPEC-TABLES.md:7135):
//
//   "Buffers are caller-owned Uint8Lists and nothing throws."
//
// THE SANITIZER. Dart has no ASan, so the memory-safety oracle is the
// language's own bounds check: every read the generated fixed-form reader
// makes is through a ByteData / Uint8List, and a read past the caller's buffer
// THROWS. "No exception escaped the reader" therefore IS "no read left the
// buffer" — a reader that lets one out has stopped refusing and has not landed
// values, and that escaping exception is the fourth answer this row says never
// happens. It is the same substitution the JS row records (its DataView throws)
// and the Elixir row records (its escaping exception class is the oracle).
//
// THE SWEEP. A valid fixed-form Chain file — the P3.schema optional side:
// string(16) name and ?Link { int32 value | min = 0, max = 1000; string(8)
// tag } — is written through the production chainFixedSave, then EVERY byte is
// flipped one bit at a time. Each mutant is read through the production
// chainFixedLoad (a fresh plan and report per mutant) and must answer exactly
// one of the three ways, never a fourth:
//
//   * report.refused != none  — a refusal BY NAME, or
//   * report.malformed        — a malformed read, or
//   * loaded == 1             — a read that LANDS VALUES.
//
// Any thrown exception, and any answer outside those three, is the fourth
// answer and a failure. The sweep counts its three answers and asserts each
// occurred, so a datapath that quietly stopped reaching the reader cannot pass.
//
// THE PRODUCTION PATH. build/tables-generated-dart/p3/, laid down by
// `make build/conformance-dart`: P3Fixed.dart's chainFixedMeasure,
// chainFixedSave, chainFixedLoad and chainFixedNewPlan, over Tblp3Fixed.dart's
// Chain/Link value classes, the known-layout table and the one read loop. The
// same emitter output the conformance driver reaches the generated tables
// through. The vector here is CONSTRUCTED from the law (the tree carries no
// fixed-form P3 bytes under testdata/conformance/tables), and the file is the
// whole file the law says to sweep: the form byte, the reserved bytes, the
// header hash, the layout length and layout bytes, and the record.
//
// THE CONTROL (recorded in RESULT.md). Breaking one constant of this law in
// the generated code — the layoutBytes bound that keeps a hostile layout
// length from walking the layout compare past the buffer — lets a mutant's
// read leave the buffer and throw, and the `escaped == 0` assertion goes RED.
// Restored, the row is GREEN again.
//
// Run from the repository root:  dart run test/conformance/dart/rows/P3.dart
// Exit 0 = green, exit 1 = red. One printed line per assertion.

import 'dart:io';
import 'dart:typed_data';

import '../../../../build/tables-generated-dart/p3/P3Fixed.dart' as p3;
import '../../../../build/tables-generated-dart/p3/Tblp3Fixed.dart'
    as fixed;

int green = 0;
int red = 0;

void check(bool ok, String what) {
  stdout.writeln('${ok ? 'ok  ' : 'FAIL'} $what');
  if (ok) {
    green++;
  } else {
    red++;
  }
}

void putText(Uint8List buffer, String value) {
  buffer.setRange(0, value.length, value.codeUnits);
}

/// Build THE LAWFUL FILE the sweep mutates: one Chain with the optional Link
/// present, so every region of the file the law names is non-trivial — a name
/// short of its bound (its slack must still be zeros), a present byte, a value
/// at 42 inside 0..1000, and a tag short of its bound.
Uint8List lawfulFile() {
  final chain = fixed.Chain();
  putText(chain.name, 'p3-row');
  chain.nameLength = 6;
  chain.linkPresent = true;
  chain.link.value = 42;
  putText(chain.link.tag, 'tag');
  chain.link.tagLength = 3;

  final bytes = Uint8List(p3.chainFixedMeasure(1));
  final wrote = p3.chainFixedSave(<fixed.Chain>[chain], 1, bytes);
  if (wrote != bytes.length) {
    stderr.writeln('FAIL: chainFixedSave wrote $wrote of ${bytes.length}');
    red++;
  }
  return bytes;
}

/// ONE MUTANT, ONE ANSWER. Reads the mutant through the production entrypoint
/// and returns one word for which of the three answers it got, or a fourth
/// description so a failure can name it.
String answer(Uint8List mutant) {
  final values = <fixed.Chain>[fixed.Chain()];
  final report = fixed.TableFixedReport();
  final plan = p3.chainFixedNewPlan();
  final loaded = p3.chainFixedLoad(
    values,
    1,
    mutant,
    mutant.length,
    plan,
    report,
  );
  if (report.refused != fixed.TableFixedRefusal.none) {
    return 'refusal:${fixed.TableFixedRefusal.name(report.refused)}';
  }
  if (report.malformed) {
    return 'malformed';
  }
  if (loaded == 1) {
    return 'landed';
  }
  return 'fourth(loaded=$loaded malformed=${report.malformed} '
      'refused=${report.refused})';
}

void main() {
  final source = lawfulFile();

  // ---- the lawful file opens and lands values (the baseline the sweep needs)
  {
    final back = <fixed.Chain>[fixed.Chain()];
    final report = fixed.TableFixedReport();
    final loaded = p3.chainFixedLoad(
      back,
      1,
      source,
      source.length,
      p3.chainFixedNewPlan(),
      report,
    );
    check(loaded == 1, 'lawful file: chainFixedLoad answers 1 record ($loaded)');
    check(
      !report.malformed && report.refused == fixed.TableFixedRefusal.none,
      'lawful file: a clean read moves no refusal and is not malformed',
    );
    check(back[0].linkPresent, 'lawful file: the optional Link is present');
    check(back[0].link.value == 42, 'lawful file: link.value lands 42');
    check(back[0].nameLength == 6, 'lawful file: nameLength lands 6');
    check(source[0] == fixed.tableFixedForm, 'lawful file: form byte 3');
  }

  // ---- THE SWEEP: every byte, one bit at a time, over the whole file
  var refusals = 0;
  var malformed = 0;
  var landed = 0;
  var escaped = 0;
  var fourth = 0;
  var firstEscape = '';
  var firstFourth = '';
  final total = source.length * 8;

  for (var i = 0; i < source.length; i++) {
    for (var b = 0; b < 8; b++) {
      final mutant = Uint8List.fromList(source);
      mutant[i] ^= 1 << b;
      try {
        final a = answer(mutant);
        if (a == 'malformed') {
          malformed++;
        } else if (a == 'landed') {
          landed++;
        } else if (a.startsWith('refusal:')) {
          refusals++;
        } else {
          fourth++;
          if (firstFourth.isEmpty) {
            firstFourth = 'byte $i bit $b: $a';
          }
        }
      } catch (e) {
        escaped++;
        if (firstEscape.isEmpty) {
          firstEscape = 'byte $i bit $b: $e';
        }
      }
    }
  }

  check(
    escaped == 0,
    'sanitizer: no single-bit mutant made a read leave the buffer — '
    '$escaped exceptions escaped a reader'
    '${firstEscape.isEmpty ? '' : ' (first: $firstEscape)'}',
  );
  check(
    fourth == 0,
    'never a fourth answer: every mutant was a refusal, malformed, or a read '
    'that lands values — $fourth otherwise'
    '${firstFourth.isEmpty ? '' : ' (first: $firstFourth)'}',
  );
  check(
    refusals + malformed + landed + escaped + fourth == total,
    'the sweep answers every mutant exactly once: '
    '${refusals + malformed + landed + escaped + fourth} of $total',
  );
  check(
    refusals > 0,
    'a refusal by name occurs under the sweep: $refusals mutants '
    '(the form byte, the header hash and the record hash)',
  );
  check(
    malformed > 0,
    'a malformed read occurs under the sweep: $malformed mutants '
    '(the reserved bytes and the layout)',
  );
  check(
    landed > 0,
    'a read that lands values occurs under the sweep: $landed mutants '
    '(the record body)',
  );

  stdout.writeln(
    'P3: ${source.length} bytes, $total single-bit mutants — '
    '$refusals refused, $malformed malformed, $landed landed, '
    '$escaped escaped — '
    '${red == 0 ? 'GREEN' : 'RED'}',
  );
  exit(red == 0 ? 0 : 1);
}

File: test/conformance/rust/rows/P3.rs
// P3 — HOSTILE BYTES, SWEEP + SANITIZER (docs/roadmap.sexp:3929, rust/P3;
// hostile-input/rust).
//
// THE LAW, docs/FIXED-FORM-ALGORITHM.md:1686 (§7 proof 5 — the card's
// mechanical pointer :1080 lands on a table separator; the governing sentence
// for the title words is :1686, verified):
//
//   "a byte-flip fuzz over the whole file, under a sanitizer, and it is not
//    optional — every offset is arithmetic over sizes a stranger wrote down,
//    so every byte, one bit at a time, is answered one of three ways and
//    never a fourth: a refusal by name, a `malformed` read, or a read that
//    lands values."
//
// The same rule at docs/SPEC-TABLES.md:7131 ("A BYTE-FLIP FUZZ OVER A WHOLE
// FILE, UNDER A SANITIZER ... ONE OF THREE WAYS AND NEVER A FOURTH ... 'The
// reader never leaves the buffer' is a claim only a sanitizer can hold"), and
// the §8 coverage-matrix row at docs/FIXED-FORM-ALGORITHM.md:1714
// ("byte-flip fuzz, sanitizer, three answers never a fourth | §7 item 5").
//
// THE PRODUCTION PATH is `chain_fixed_load`, the generated form-3 reader for
// build/tables-generated-rust/tblp3 (test/tables/P3.schema — table Chain with
// its ?Link optional), reached the way row C10 reaches FE1: one module per
// generated source file by #[path], skipping the WIRE module (`p3.rs`)
// because the fixed modules have no dependency on the serialize runtime.
// The unit and fixture are the ones the sibling cell rows already drive:
// test/conformance/go/rows/P3_test.go and test/conformance/dart/rows/P3.dart.
//
// THE VECTOR, from the law's own datum: `chain_fixed_save` writes one lawful
// WHOLE file — 16 header bytes + 4 block-length + 106 layout + one 45-byte
// record = 171 bytes — fixture values mirrored from the go/dart rows (name
// "tips" length 4, link present with value 42, tag "tag" length 3). The tree
// carries no form-3 whole-file fixture under testdata/ (that corpus is
// build/fixedform-corpus, a make target), so the file is written by the
// generated writer itself; every offset of it is then arithmetic the reader
// takes from stranger-written sizes.
//
// THE SANITIZER HALF is the leg's own reading of the clause
// (test/rust-fixedform/src/main.rs:2295-2305): rustc stable carries no ASan,
// the BUILD is the sanitizer — every index this reader reaches runs through
// checked slices and `.get()`, and any escape out of the buffer is a PANIC,
// which under `rustc --test` fails this binary. A panic inside the sweep IS a
// fourth answer and never arrives quietly.
//
// No assertion of this cell existed under test/conformance/rust/ (STEP 2);
// the leg's earlier twin over fx1.bin is the_hostile_sweep at
// test/rust-fixedform/src/main.rs:2313, run by `make tables-rust-fixedform`
// (make/rust.mk:277), CI .github/workflows/ci-fast.yml:616. This file is the
// matrix row for the cell. One printed line per assertion.
//
// Run from ./repo (no --extern/-L needed: the included fixed modules have no
// dependency on the conformance crate's serialize runtime, like C10 and E6):
//
//   rustc --edition 2021 --test test/conformance/rust/rows/P3.rs \
//     -o build/rows-rust-P3 && ./build/rows-rust-P3
//
// Exit 0 = green, non-zero = red.
#![allow(unused_imports)]

#[path = "../../../../build/tables-generated-rust/tblp3/src/fixed_runtime.rs"]
mod fixed_runtime;
#[path = "../../../../build/tables-generated-rust/tblp3/src/p3_records.rs"]
mod p3_records;
#[path = "../../../../build/tables-generated-rust/tblp3/src/p3_fixed.rs"]
mod p3_fixed;

pub use fixed_runtime::*;
pub use p3_records::*;
pub use p3_fixed::*;

// One lawful whole file, by the generated writer: header, layout, one record.
fn lawful_file() -> Vec<u8> {
    let row = ChainRow {
        name: {
            let mut name = [0u8; 17];
            name[..4].copy_from_slice(b"tips");
            name
        },
        name_length: 4,
        link: LinkRow {
            value: 42,
            tag: {
                let mut tag = [0u8; 9];
                tag[..3].copy_from_slice(b"tag");
                tag
            },
            tag_length: 3,
        },
        link_present: true,
    };
    let mut file = vec![0u8; chain_fixed_measure(1)];
    let wrote = chain_fixed_save(&[row], &mut file);
    assert_eq!(
        wrote,
        Some(file.len()),
        "the fixture must write a whole file: {wrote:?}"
    );
    file
}

fn load(
    data: &[u8],
) -> (
    Option<usize>,
    [ChainRow; 4],
    TableFixedReport,
) {
    let mut values = [ChainRow::default(); 4];
    let mut plan = [TableFixedEntry::default(); 64];
    let mut remap = [0u16; 64];
    let mut report = TableFixedReport::default();
    let n = chain_fixed_load(&mut values, data, &mut plan, &mut remap, &mut report);
    (n, values, report)
}

// THE POSITIVE CONTROL, IN THE SAME BINARY: the lawful file lands its values
// and moves nothing, so every refusal below is the forgery and not a reader
// that answers hostile to everything.
#[test]
fn p3_a_lawful_file_lands_values() {
    println!("P3 positive control: the lawful Chain file lands its values and counts nothing");
    let (n, values, report) = load(&lawful_file());
    assert_eq!(n, Some(1), "the lawful file reads one record: {report:?}");
    assert_eq!(
        report,
        TableFixedReport::default(),
        "a lawful read moves no counter and sets no flag: {report:?}"
    );
    assert_eq!(&values[0].name[..4], b"tips", "name lands");
    assert_eq!(values[0].name_length, 4, "name_length lands");
    assert!(values[0].link_present, "link present lands");
    assert_eq!(values[0].link.value, 42, "link value lands");
    assert_eq!(&values[0].link.tag[..3], b"tag", "tag lands");
}

// THE CELL — every byte of the whole file, one bit at a time (§7 proof 5):
// each mutant is answered one of three ways — a refusal by name, a
// `malformed` read, or a read that lands values — and never a fourth. A panic
// anywhere in the sweep (the sanitizer's verdict: a read that left the buffer)
// fails this test before the counts are checked.
#[test]
fn p3_every_single_bit_mutation_gets_one_of_three_answers_never_a_fourth() {
    let file = lawful_file();
    let mut refused: u32 = 0;
    let mut damaged: u32 = 0;
    let mut landed: u32 = 0;
    let mut fourth: Option<String> = None;
    for at in 0..file.len() {
        for bit in 0..8u32 {
            let mut hit = file.clone();
            hit[at] ^= 1u8 << bit;
            let (n, _, report) = load(&hit);
            // THE THREE ANSWERS, classified exactly as the leg's twin does
            // (test/rust-fixedform/src/main.rs:2347-2354).
            let named = n.is_none()
                && report.refused
                && report.reason != TableFixedReason::None
                && !report.malformed;
            let hurt = report.malformed && !report.refused;
            let clean = n.is_some() && !report.refused && !report.malformed;
            if named {
                refused += 1;
            } else if hurt {
                damaged += 1;
            } else if clean {
                landed += 1;
            } else if fourth.is_none() {
                fourth = Some(format!(
                    "byte {at} bit {bit}: n={n:?} refused={} reason={} malformed={} — \
                     a refusal, a malformed read or a read that lands values, never a fourth",
                    report.refused,
                    report.reason.name(),
                    report.malformed
                ));
            }
        }
    }
    let total = (file.len() * 8) as u32;
    if let Some(what) = fourth {
        panic!("{what}");
    }
    assert_eq!(
        refused + damaged + landed,
        total,
        "every mutation answered exactly one way: refused={refused} malformed={damaged} landed={landed}"
    );
    println!(
        "P3: {total} single-bit mutants over {} bytes — refused={refused} malformed={damaged} landed={landed}, three answers never a fourth",
        file.len()
    );
}

// ANTI-VACUOUS: all three answers FIRED on this fixture — a sweep that only
// ever landed (a reader that stopped refusing) or only ever refused (a reader
// that decodes nothing) would still have answered "never a fourth".
#[test]
fn p3_all_three_answers_fire() {
    let file = lawful_file();
    let mut refused: u32 = 0;
    let mut damaged: u32 = 0;
    let mut landed: u32 = 0;
    for at in 0..file.len() {
        for bit in 0..8u32 {
            let mut hit = file.clone();
            hit[at] ^= 1u8 << bit;
            let (n, _, report) = load(&hit);
            if n.is_none() && report.refused && report.reason != TableFixedReason::None && !report.malformed {
                refused += 1;
            } else if report.malformed && !report.refused {
                damaged += 1;
            } else if n.is_some() && !report.refused && !report.malformed {
                landed += 1;
            }
        }
    }
    println!(
        "P3: all three answers fire on the sweep — refused={refused} malformed={damaged} landed={landed}"
    );
    assert!(refused > 0, "a refusal by name must fire: {refused}");
    assert!(damaged > 0, "a malformed read must fire: {damaged}");
    assert!(landed > 0, "a read that lands values must fire: {landed}");
}

// THE LENGTH IS STRANGER-CONTROLLED TOO: one byte off the end is arithmetic
// over a size a stranger wrote down — DAMAGE by name (`malformed`), never a
// silent short read and never a refusal dressed as one.
#[test]
fn p3_a_one_byte_short_file_is_damage() {
    println!("P3: a one-byte-short file is DAMAGE, not a short read");
    let file = lawful_file();
    let mut short = file.clone();
    short.pop();
    let (n, _, report) = load(&short);
    assert_eq!(n, None, "the short file does not land: {report:?}");
    assert!(report.malformed, "one byte off the end is malformed: {report:?}");
    assert!(
        !report.refused,
        "a ragged tail is damage, not a refusal: {report:?}"
    );
    assert_eq!(
        report.reason,
        TableFixedReason::None,
        "and it carries no refusal name: {report:?}"
    );
}
